agent/api/task/task.go

```diff
@@ -3240,7 +3240,7 @@ func (task *Task) initializeEnvfilesResource(config *config.Config, credentialsM
 				return errors.Wrapf(err, "unable to initialize envfiles resource for container %s", container.Name)
 			}
 			task.AddResource(envFiles.ResourceName, envfileResource)
-			container.BuildResourceDependency(envfileResource.GetName(), resourcestatus.ResourceCreated, apicontainerstatus.ContainerCreated)
+			container.BuildResourceDependency(envfileResource.GetName(), resourcestatus.ResourceStatus(envFiles.EnvFileCreated), apicontainerstatus.ContainerCreated)
```

Comment:
Could you help me understand why do we want to use `resourcestatus.ResourceStatus(envFiles.EnvFileCreated)`

Reply:
Sure thing! I believe it to be a better practice to build the dependency on the status defined by the specific type of task resource, rather than the status defined for task resources in general. Consider the following: what if in the future we want to define a new status specific to environment file resource?

We also see this practice being employed already for most other types of task resources:
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L484
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L712
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1045
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1083
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1090
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1151
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1158
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task/task.go#L1481

Reply:
Also observe that `resourcestatus.ResourceStatus(EnvFileCreated)` is the one used in `envfile.go`:

https://github.com/aws/amazon-ecs-agent/blob/master/agent/taskresource/envFiles/envfile.go#L136
https://github.com/aws/amazon-ecs-agent/blob/master/agent/taskresource/envFiles/envfile.go#L223
https://github.com/aws/amazon-ecs-agent/blob/master/agent/taskresource/envFiles/envfile.go#L249

and that there are no references to `resourcestatus.ResourceCreated` in this file. The one used in `task.go` should match what is being used in `envfile.go`.

Reply:
Please let me know if I'm missing anything - looking at `resourcestatus.ResourceStatus(envFiles.EnvFileCreated)`, if we end up having more statuses for envFile resource (10 let's say), we could very well end up with an invalid `ResourceStatus` as the iota goes out of bound. The reason why we've been able to do it for all resources without an issue is because, I believe, all resource-specific statuses are actually **identical** to the generic statuses hence 1:1 mapping (that's the case when I spot-checked asm, volume, and firelens - probably the others as well). 

This is getting out of scope of the problem we are trying to solve at hand. But I feel like we might want to revisit how we are using a generic `ResourceStatus` alongside the status defined for a given type of resource (e.g. `envFiles.EnvFileCreated`). I'm not actually sure why we have both. When we drive a task/container from the top-level state machine, we can only use one status, so it's an either/or. If we want to do respect each resource's own statuses and transitions, then the `resourcestatus.ResourceStatus` should've just been an int type, not iota constant. Again, totally out of scope and probably should be followed up separately. But given what we currently have, I don't think this "conversion" here actually makes sense.

Reply:
Thanks for calling this out. Agreed this is getting out of scope of the change we are making in this pull request, but I will definitely create an internal backlog item in the ECS Agent team to revisit and investigate this as it is a bit confusing/unintuitive to figure out what exactly we are trying to achieve in the way we currently use these statuses.